core/shared/src/main/scala/zio/stm/TArray.scala

```diff
@@ -16,139 +16,193 @@
+   */
   final def collectFirstM[E, B](pf: PartialFunction[A, STM[E, B]]): STM[E, Option[B]] =
     find(pf.isDefinedAt).flatMap {
       case Some(a) => pf(a).map(Some(_))
       case _       => STM.succeed(None)
     }
 
-  /** Determine if the array contains a specified value. */
+  /**
+   * Determine if the array contains a specified value.
+   */
   final def contains(a: A): STM[Nothing, Boolean] = exists(_ == a)
 
-  /** Count the values in the array matching a predicate. */
-  final def count(predicate: A => Boolean): STM[Nothing, Int] = fold(0) { (n, a) =>
-    if (predicate(a)) n + 1 else n
-  }
+  /**
+   * Count the values in the array matching a predicate.
+   */
+  final def count(p: A => Boolean): STM[Nothing, Int] =
+    fold(0)((n, a) => if (p(a)) n + 1 else n)
 
-  /** Count the values in the array matching an STM predicate. */
-  final def countM[E](predicate: A => STM[E, Boolean]): STM[E, Int] = foldM[E, Int](0) { (n, a) =>
-    predicate(a).map(result => if (result) n + 1 else n)
-  }
+  /**
+   * Count the values in the array matching an STM predicate.
+   */
+  final def countM[E](p: A => STM[E, Boolean]): STM[E, Int] =
+    foldM(0)((n, a) => p(a).map(result => if (result) n + 1 else n))
 
-  /** Determine if the array contains a value satisfying a predicate. */
-  final def exists(predicate: A => Boolean): STM[Nothing, Boolean] = find(predicate).map(_.isDefined)
+  /**
+   * Determine if the array contains a value satisfying a predicate.
+   */
+  final def exists(p: A => Boolean): STM[Nothing, Boolean] = find(p).map(_.isDefined)
 
-  /** Determine if the array contains a value satisfying an STM predicate. */
-  final def existsM[E](predicate: A => STM[E, Boolean]): STM[E, Boolean] =
-    countM(predicate).map(_ > 0) //Existence should not be order dependent so the entire array needs to be observed.
+  /**
+   * Determine if the array contains a value satisfying an STM predicate.
+   */
+  final def existsM[E](p: A => STM[E, Boolean]): STM[E, Boolean] =
+    countM(p).map(_ > 0)
 
-  /** Find the first element in the array matching a predicate. */
-  final def find(predicate: A => Boolean): STM[Nothing, Option[A]] =
+  /**
+   * Find the first element in the array matching a predicate.
+   */
+  final def find(p: A => Boolean): STM[Nothing, Option[A]] =
     if (array.isEmpty) STM.succeed(None)
     else
       array.head.get.flatMap { a =>
-        if (predicate(a)) STM.succeed(Some(a))
-        else new TArray(array.tail).find(predicate)
+        if (p(a)) STM.succeed(Some(a))
+        else new TArray(array.tail).find(p)
       }
 
-  /** Find the last element in the array matching a predicate. */
-  final def findLast(predicate: A => Boolean): STM[Nothing, Option[A]] =
-    new TArray(array.reverse).find(predicate)
+  /**
+   * Find the last element in the array matching a predicate.
+   */
+  final def findLast(p: A => Boolean): STM[Nothing, Option[A]] =
+    new TArray(array.reverse).find(p)
```

Comment:
This is counter-intuitive, you would expect subsequent operations to be done on the same array, but here a new one is created, which could yield different results on those operations.

Reply:
This one's actually the old one (diff due to renaming). Technically, it is on the same one (TArray being value class), however, I have similar concerns on a couple more methods (e.g. `existsM`).

Reply:
Edit: I've re-read what you wrote, and yes, this is even more extreme actually.